### ISSUE TITLE
keyboard_layout: Removal of setxkbmap/xset method

### ISSUE THREAD

**Issue by saengowp** [ultrabug/py3status#1816]
In `keyboard_layout` module, There are 2 method to control the keyboard layout

1. `setxkbmap`/`xset`
2. `xkblayout-state`

The `setxkbmap`/`xset` method relies on keyboard LED and overwriting layout. This causes following issue:

- `py3status` became responsible for keyboard layout.
- Other application relying on `XKB` won't work properly (eg. Layout switching hotkey)
- keyboard variant is incorrectly displayed

deprecating this method would resulted in:

- User would no longer be able to configure their keyboard using this module.
- Module no longer work for user who doesn't has `xkblayout-state`

I believe that we should deprecate this method then notify user to install `xkblayout-state` and setup their keyboard properly. this would fixes all mentioned issue.

**Comment by lasers** [comment]
PRs welcome. It might be easier to make this from scratch. It might be nice to support more placeholders too, i.e. (`{v}`, `{V}`, `{c}`, `{s}`, etc in xkblayout-state).  We got some old compatibility there too.

I wonder about making classes if there are more than just `setxkbmap`, `xkblayout` in case users does not have `xkblayout` initially installed... to reduce/eliminate throwing py3 errors about missing software.

**Comment by lasers** [comment]
be327773b6d6cf60a0fd6121d0f0f13bf580af55 cquses qn issue too::: zith pqrsing lqyout betzeen zorking `%S` qnd broken `%E`.:  öß====D°°°

**Comment by saengowp** [comment]
Okay, I will submit PRs implementing above proposal. That might take some time.

**Comment by lasers** [comment]
Not finished. Going to bed. Gnight. https://github.com/lasers/py3status/tree/newmodule-keyboard

**Comment by saengowp** [comment]
Oh, I didn't notice that you've already working on it. Thank you!

**Comment by lasers** [comment]
@saengowp I'm currently stuck on `Setxkbmap` class. Seems like we can't specify multiple `-layout`, but we can do that with `-variant` so it's not something straightforward to put together for `setup()` + other functions... so I have some questions.

Is it (technically) better to use `setxkbmap` + https://github.com/ultrabug/py3status/pull/1183 without third parties tools?

```python
# create keyboard_layout switching
order += "switch keyboard_layout"
switch keyboard_layout {
    button_next = 4
    button_previous = 5
    cache_timeout = 10
    list = [
        ('\?color=#268BD2 fr', 'setxkbmap fr'),
        ('\?color=#F75252 ru', 'setxkbmap ru'),
        ('\?color=#FCE94F ua', 'setxkbmap ua'),
        ('\?color=#729FCF us', 'setxkbmap us'),
    ]
}
```
With this, users could specify variants, rules, etc too...

I take it we might be able to avoid those third parties tools if we could utilize `setxkbmap` for everything instead? But what about...
>Other application relying on XKB won't work properly (eg. Layout switching hotkey)

Can you explain more? The other issue is that the `-variant` can have single value for all layouts or comma-separated for each layout... maybe we need to consider this in the code too.... is which why I wonder about dropping classes to make a pure `setxkbmap` module.

**Comment by saengowp** [comment]
Sorry if this is a bit long and doesn't answer every question.

> Other application relying on XKB won't work properly (eg. Layout switching hotkey)

Here's my investigation into `setxkbmap`.

XKB provides an API for retrieving current selected layout and for changing the selection. However, `setxkbmap` doesn't provide access to this API so we have to do a *weird thing* if we were to rely on `setxkbmap`.

**My attempt at *Weird thing***
1. We can forcibly replace all layout with a single layout we want eg. `setxkbmap ru` but there's a problem!
     - Other application (Maybe some other layout modifying app) will think that user only has one layout.
     - (Corollary) Keyboard switching shortcut no longer work because now it only knew one layout.

2. Maybe we can put selected layout in front of every other layout to preserve layout list but there's a catch!
     - `setxkbmap -option grp:*_toggle` remember selected layout by index, if the user used this option then switch layout using both shortcut and this module, selected layout will not be what we think it is.

3. Maybe we rotate the list.
     - Some other application will not notice selection change at all (eg. `xkb-switch -W`)

Additionally, we can't determine selected layout even with `setxkbmap -verbose 10`. So I gave up on `setxkbmap` and let other 3rd party app call those C API for me.

In the end, I think `setxkbmap` is too weak for this task.

**Unrelated note:** I think your rewrite with class is going to make `sway` support very easy.

**Comment by lasers** [comment]
Thoughts on dropping `setxkbmap` class and rename `keyboard` to `xkb`? Maybe add a note/link for users wanting to set up their `setxkbmap` since that seems like out of scope for `xkb` module because we potentially would have to add config(s) for layouts, variants, rules?, options?, etc.

Hand-off approach may be better and we would only deal with xkb switchers. Should solve this issue too.

**Comment by saengowp** [comment]
I agree with your solution.

**Comment by lasers** [comment]
@saengowp Done. Can you try https://github.com/ultrabug/py3status/tree/newmodule-xkb? 
@ultrabug @maximbaz Comments? Thoughts? Feedback? Alternative solutions? Etc.

**Comment by maximbaz** [comment]
Hey guys, I like that you found tools that make layout switching work better, but I'd like to question the need for deprecating `setxkbmap` — for example, I don't use variants, and in general I use this module purely to query the current layout (I have `button_prev = button_next = None`), and for my purposes `setxkbmap` seems to work quite well. 

I also noticed that `xkb-switch` and `xkblayout-state` are currently not present in Arch Linux official repos, in case of Arch it's not a problem to add them, but this tells me that these tools are probably missing in other distros as well, and you might cause unnecessary pain of forcing a lot of people to find ways to install them, who won't really benefit of changing from `setxkbmap` to these fancy tools 🙂

**Comment by saengowp** [comment]
- @lasers The new modules is working properly for me and I see no problem. (On sway there's a problem but it's upstream anyway)
- @maximbaz I just figured out that it's possible to not use 3rd party tools ([Proof of concept](https://gist.github.com/saengowp/49b677869b034208fe7731352514f9d1)) but it look horrifying to me.

**Comment by lasers** [comment]
@maximbaz That's why I made `xkb` to leave others untouched/"unaffected". Imho, Module `xkb` is the nice way to get current layout as `setxkbmap` does not offer this... and hadn't been worked / updated in long time so this feature is sorely missing. Also, there may be a bug in `keyboard_layout` because it's trying to deal with variant too... and offering users to configure their layouts with `layouts` instead of using `setxkbmap`. 

@saengowp Indeed, it looks terrible. Is there a way to get symbols instead of names? If yes, it could be good enough for `read-only` xkb with no scrolling... what @maximbaz want to do here.

**Comment by lasers** [comment]
@maximbaz 

>in general I use this module purely to query the current layout...

`setxkbmap -query` prints all layouts regardless... and does not offer a way to query current layout.

>and for my purposes setxkbmap seems to work quite well.

The code say this:

1) In case of having single layout, this should work 100% fine.
2) In case of having two or more layouts, it queries `xset -q` too... then works with `LED mask:  00000000` to figure out if you're using a different layout. If you are, there are no guarantee that it will return the correct layout... so it can break here depending on the layouts. If you only have two layouts or use common layouts, then it can work fine too (i.e. lucky).

https://github.com/ultrabug/py3status/blob/86b62726303ec7accbdd1cf86d389c16c80ed757/py3status/modules/keyboard_layout.py#L146-L160

**Comment by maximbaz** [comment]
Fair enough, if that solution is technically awful and is working only in rare cases, yes lets get rid of it. I don't mind packaging `xkb-switch` and/or `xkblayout-state` for Arch to simplify the upgrade for people (the former looks more alive, should probably be preferred). 

A couple of extra thoughts:

- Should we include this in `4.0`, since it's a breaking change?
- I think we shouldn't include `X` in the name of the module (as in `xkb`), because hopefully we can find another tool and support Wayland in the future. The name `keyboard_layout` is fine I think.

**Comment by lasers** [comment]
>if that solution is technically awful and is working only in rare cases,

Technically awful, but should work okay for most users with two layouts, Also, this L147 comment.
```
# this method works only for the first two predefined layouts.
```

I *think* `xkblayout-state` is better than `xkb-switch`, but we can support both. I added support for `python-xkbgroup` too cause why not...

>Should we include this in 4.0, since it's a breaking change?
>I think we shouldn't include X in the name of the module (as in xkb), because hopefully we can find another tool and support Wayland in the future. The name keyboard_layout is fine I think.

I leave this to @ultrabug. I'm not using `keyboard_layout` or `xkb` so I haz no preference. The trend, I think, is to add a new module and gracefully deprecate an old module sometimes in future.
* ~~`nvidia_temp`~~ `nvidia_smi` (I want to)
* ~~`scratchpad_async`~~ `scratchpad` (I want to)
* ~~`scratchpad_counter`~~ `scratchpad` (I want to)
* ~~`weather_yahoo`~~ `weather_owm` (Done)
* ~~`window_title_async`~~ `window_title` (I want to. Need to code first.)
* ~~`window_title`~~ `window_title` (I want to. Need to code first.)

@saengowp Can you confirm `python-xkbgroup` is working okay for you? I figure I should add this before somebody else try to... and I changed from default `{s}` to `{e}` to stay consistency between classes.

**Comment by saengowp** [comment]
Sorry for late reply @lasers .

I've tested the your new `xkb` module and found that it work perfectly in all test condition.

I've slight objection regarding order of command, I think that 'xkb-switch' should be last command since it has the least capability, Other that that, I think this module is finished.

**Comment by lasers** [comment]
>I think that xkb-switch should be last command since it has the least capability
>(...I think order depends on your discretion then)

Sure. It was sorted alphabetically. Now, we go by parity. `xkblayout-state` == `xkbgroup`, then `xkb-switch`.

>I think we shouldn't include X in the name of the module (as in xkb), because hopefully we can find another tool and support Wayland in the future. The name `keyboard_layout` is fine I think.

I put out a PR now. Still @ultrabug's decision. I just go with `xkb` because people who uses multiple layouts would know what `xkb` mean.

If/when X/Wayland have a name to replace `xkb` with, then we could change the module name too. I just wanted to mention that all of our binaries starts with `xkb` too.

If we want to use `keyboard_layout` instead and break configs, fine with me too. Let me know.

**Comment by ultrabug** [comment]
So I never used this module so it is hard for me to say what's best here from a user's perspective.

If we were to merge those modules, I would prefer to do it when going 4.0 tbh which is an expected breakage at some point.

So the question gets down to: do we need to add another module to py3status? If it's more powerful and goes on the sway path, why not!

So now I'll review the new module proposal if that's okay

**Comment by lasers** [comment]
@ultrabug I don't know how much you looked into this issue, but I'll explain the issue a bit more.

>So the question gets down to: do we need to add another module to py3status? If it's more powerful and goes on the sway path, why not!

Can replace `keyboard_layout` in favor of `xkb` to avoid add another module... like `weather_yahoo`.

-----

The issue...

`xkb` is not more powerful, but it avoids doing the wrong thing when users don't have `xkblayout-state` installed... From what I gather, the users should run `setxkbmap` once somewhere in `xinitrc` or `i3 config`, then just switch them around...

In `keyboard_layout`, it uses `layouts: specify a list of layouts to use (default None)` to run `setxkbmap` repeatedly to issue single language layouts instead, i.e...

Wrong (setxkbmap)
```
(command 1) setxkbmap -layout "us"   # us only
(command 2) setxkbmap -layout "fr"   # fr only
(command 3) setxkbmap -layout "ru"   # ru only

(xkb switch) xkblayout-state set +1  # won't work... still on ru
(xkb switch) xkblayout-state set +1  # won't work... still on ru
(xkb switch) xkblayout-state set +1  # won't work... still on ru
(xkb switch) xkblayout-state set +1  # won't work... still on ru
```
Right (xkblayout-state... and xkbgroup, xkb-switch)
```
(command 1) setxkbmap -layout "us,fr,ru"

(xkb switch) xkblayout-state set +1  # works... now fr
(xkb switch) xkblayout-state set +1  # works... now ru
(xkb switch) xkblayout-state set +1  # works... now us
(xkb switch) xkblayout-state set +1  # works... now fr
```
Basically, we're using `setxkbmap` to set languages instead of switching languages.

**Comment by lasers** [comment]
Closed via #1817.
